BasicSensorsApiKotlin/app/src/main/java/com/google/android/gms/fit/samples/basicsensorsapikotlin/MainActivity.kt

```diff
@@ -0,0 +1,343 @@
+
+const val TAG = "BasicSensorsApi"
+
+/**
+ * This sample demonstrates how to use the Sensors API of the Google Fit platform to find available
+ * data sources and to register/unregister listeners to those sources. It also demonstrates how to
+ * authenticate a user with Google Play Services.
+ */
+class MainActivity : AppCompatActivity() {
+    private val fitnessOptions = FitnessOptions.builder().addDataType(DataType.TYPE_LOCATION_SAMPLE).build()
+
+    // Two maps are used in conjunction with the OAuth flow. This is in order to identify what
+    // function to execute once the flow has successfully completed.
+    private val signInActionMap = mapOf<() -> Any, Int>(
+            ::findFitnessDataSources to 1
+    )
+    private val signInResponseMap = signInActionMap.entries.associate { (k, v) -> v to k }
+
+    // [START dataPointListener_variable_reference]
+    // Need to hold a reference to this listener, as it's passed into the "unregister"
+    // method in order to stop all sensors from sending data to this listener.
+    private var dataPointListener: OnDataPointListener? = null
+
+    // [END dataPointListener_variable_reference]
+
+    // [START auth_oncreate_setup]
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        // Put application specific code here.
+        setContentView(R.layout.activity_main)
+        // This method sets up our custom logger, which will print all log messages to the device
+        // screen, as well as to adb logcat.
+        initializeLogging()
+        // When permissions are revoked the app is restarted so onCreate is sufficient to check for
+        // permissions core to the Activity's functionality.
+        checkPermissionsAndRun(::findFitnessDataSources)
+    }
+
+    private fun checkPermissionsAndRun(signInAction: () -> Any) {
+        if (hasRuntimePermissions()) {
+            fitSignIn(signInAction)
+        } else {
+            requestRuntimePermissions(signInAction)
+        }
+    }
+
+    /**
+     * Checks that the user is signed in, and if so, executes the specified function. If the user is
+     * not signed in, initiates the sign in flow, specifying the post-sign in function to execute.
+     *
+     * @param signInAction The function to execute if already signed in, or once signed in.
+     */
+    private fun fitSignIn(signInAction: () -> Any) {
+        if (!hasOAuthPermissions()) {
+            val requestCode = signInActionMap[signInAction]
+            requestCode?.let {
+                GoogleSignIn.requestPermissions(
+                        this,
+                        requestCode,
+                        getGoogleAccount(), fitnessOptions)
+            }
+        } else signInAction()
+    }
+
+    private fun hasOAuthPermissions() = GoogleSignIn.hasPermissions(getGoogleAccount(), fitnessOptions)
+
+    private fun getGoogleAccount() = GoogleSignIn.getAccountForExtension(this, fitnessOptions)
```

Comment:
Comment for getAccountForExtension()... not sure if it's different than just an getGoogleAccount() method on GoogleSIgnIn... also, I don't know much about sign in so I might be missing something obvious.